### ISSUE TITLE
Regression predictions remain nearly constant around 5.8 after fine-tuning on an AChE pIC50 dataset

### ISSUE THREAD

**Issue by n52q9kqb58-a11y** [TomAIDD/NaFM-Official#5]
Hello,

Thank you for developing and sharing NaFM.
I am trying to use the model for target-specific virtual screening. My goal is to fine-tune NaFM using known activity data for a specific target and then predict and rank the activities of additional compounds.
In my current experiment, the target is AChE. I collected compound activity data from ChEMBL, including SMILES strings and pIC50 values. I selected approximately 2,000 compounds for regression fine-tuning.
Fine-tuning command

I ran the following command:

python train.py \
  --task finetune \
  --num-epochs 300 \
  --emb-dim 1024 \
  --num-layer 6 \
  --drop-ratio 0.1 \
  --dataset Regression \
  --dataset-root downstream_data/AChE \
  --pretrained-path NaFM.ckpt \
  --lr 5.0e-5 \
  --lr-min 1.0e-5 \
  --batch-size 128 \
  --save-interval 5 \
  --log-dir ./wd \
  --early-stopping-patience 50 \
  --dataset-arg ACHE \
  --seed 0

The training data are stored in:
downstream_data/AChE/regression_data.csv
The dataset contains molecular SMILES strings and their corresponding experimental pIC50 values.

Inference command
After training, I used the generated checkpoint for regression inference:
python inference.py \
  --task regression \
  --downstream-data ./smiles.csv \
  --checkpoint-path ./wd/last.ckpt \
  --num-classes 1

Unexpected result
The inference procedure completed successfully, but nearly all predicted values were concentrated around 5.8, with very little variation among structurally different compounds.
Representative results from predictions.csv are shown below:
SMILES Prediction COc1cc2c(cc1OC)CN(CC1CCN(Cc3cccc(/C=N/O)c3O)CC1)CC2 [5.791465] COc1cc2c(cc1OC)C(c1ccccc1)N(CC1CCN(Cc3cc(C)cc(/C=N/O)c3O)CC1)CC2 [5.7891154] COc1cccc(CN(C)[C@@H](C)[C@H]2[C@H](O)C[C@@]3(C)[C@@H]4CC[C@H]5[C@]6(C)CSC(C(C)C)=N[C@H]6CC[C@@]56C[C@@]46C(=O)C[C@]23C)c1 [5.8067455] CC(C)C1=N[C@H]2CC[C@]34C[C@]35C(=O)C[C@]3(C)[C@@H]([C@H](C)N(C)Cc6cccc(C(F)(F)F)c6)[C@H](O)C[C@@]3(C)[C@@H]5CC[C@H]4[C@]2(C)CS1 [5.810594] COc1ccc(CNc2c3c(nc4cc(Cl)ccc24)CCCC3)cc1OC [5.8027673] O=C(CCN1CCN(Cc2ccccc2)CC1)Nc1ccc2nc3n(c(=O)c2c1)CCC3 [5.789249] C#CC(C)NCCc1cccc(OC(=O)N(C)C)c1 [5.761987] Clc1ccc2c(NCc3ccccn3)c3c(nc2c1)CCCC3 [5.8131413]

Although the molecules are structurally diverse, the predictions span only a very narrow range. This appears similar to a model predicting approximately the mean value of the training labels rather than learning a meaningful structure–activity relationship.

Could you please help me determine which step or parameter in my workflow may be incorrect?
I provide the complete regression_data.csv, predictions.csv and checkpoint information.
Thank you very much for your time and assistance.

[predictions.csv](https://github.com/user-attachments/files/29238441/predictions.csv)
[regression_data.csv](https://github.com/user-attachments/files/29238447/regression_data.csv)
[metrics.csv](https://github.com/user-attachments/files/29241028/metrics.csv)

**Comment by TomAIDD** [comment]
Hi,

Thank you very much for your detailed report and for sharing the commands and files.

This is a useful observation, because it touches on an important distinction between direct pIC50 regression and target-specific virtual screening. In direct regression, the raw numerical outputs are optimized for point-wise error metrics such as MSE/RMSD, whereas in virtual screening the practical objective is usually to prioritize and enrich active compounds at the top of a ranked list. Therefore, a regression model may produce a relatively compressed numerical range while still retaining some activity-related ranking signal. However, if the goal is prospective screening, we generally recommend evaluating and using ranking-oriented metrics/protocols rather than relying on raw pIC50 regression values.

From your description, the training/inference procedure itself seems to have run successfully. A few implementation details are worth checking first:

1. Please make sure the regression data are placed under:

   `downstream_data/AChE/raw/regression_data.csv`

   In the current implementation, the Regression dataset reads 'raw/regression_data.csv', filters rows by: ‘target_id == NPT{dataset_arg}’

   and uses the column 'activity_value' as the regression label. 

2. If you modified 'regression_data.csv' after a previous run, please delete the cached processed file, for example:

    `downstream_data/AChE/processed/NPC_ACHE.pt`

   Otherwise PyTorch Geometric may continue loading the previously processed dataset.

3. For virtual screening, we recommend using the screening/classification protocol rather than direct pIC50 regression. In our screening experiments, compounds are converted into binary active/inactive labels according to an activity threshold, and the model is then used to rank candidate molecules by the positive-class score.

For example, for AChE you may define:

```
    pIC50 >= 6  -> active, label = 1
    pIC50 < 6   -> inactive, label = 0
```

or use pIC50 >= 5 if you want a more inclusive threshold. The better threshold depends on your desired hit definition and the distribution of your ChEMBL data.

The corresponding data format should follow the External/screening setup, with columns such as:

`    SMILES,target_id,activity_label,fold,is_coconut`

where 'activity_label' is 0/1. After fine-tuning a binary model, you can use the 'screen' inference mode to rank unlabeled or candidate compounds by their positive-class score.

So in short: the regression output you observed is not necessarily due to a runtime error, but direct regression may not be the best setting for virtual screening. For ranking candidates, we suggest converting the task into active/inactive prediction and evaluating by enrichment/ranking metrics such as EF1%, EF5%, EF10%, AUPRC, or top-k hit recovery.

Best,
Yuheng Ding
